doc/designs/thread-api.md

```diff
@@ -0,0 +1,304 @@
+ */
+int OSSL_thread_pool_enter(OSSL_THREAD_POOL_COOKIE *cookie);
+
+/*
+ * Causes any in progress calls to OSSL_thread_pool_enter, or any future such
+ * call made using the given cookie, to return.
+ *
+ * Returns 0 if the cookie is invalid. Otherwise returns 1. There is no
+ * indication of whether the call caused any OSSL_thread_pool_enter call to
+ * return.
+ *
+ * Multiple calls to this function (before any subsequent call to
+ * OSSL_THREAD_POOL_COOKIE_reset) are idempotent.
+ */
+int OSSL_thread_pool_exit(OSSL_THREAD_POOL_COOKIE *cookie);
+
+/*
+ * This is like OSSL_thread_pool_exit, but blocks until all calls to
+ * OSSL_thread_pool_enter have returned. It may also be called after any call to
+ * OSSL_thread_pool_exit to wait until all calls to OSSL_thread_pool_enter have
+ * returned.
+ */
+int OSSL_thread_pool_exit_wait(OSSL_THREAD_POOL_COOKIE *cookie);
+```
+
+Manual threads count towards the thread limit for the purposes of the default
+thread model. However, the caller is not prevented from donating threads
+manually which exceed the thread limit. This means, for example, you can leave
+the thread limit set at its default of zero, meaning that OpenSSL is not allowed
+to create threads, and donate threads to it using this API.
+
+If the default model thread limit were set to e.g. 4, and you donated two
+threads using the manual API, OpenSSL would only create up to two threads which
+it manages itself.
+
+Capability Detection
+--------------------
+
+These functions allow the caller to determine if OpenSSL was built with threads
+support.
+
+```c
+/*
+ * Retrieves flags indicating what threading functionality OpenSSL can support
+ * based on how it was built and the platform on which it was running.
+ */
+/* Is thread pool functionality supported at all? */
+#define OSSL_THREAD_SUPPORT_FLAG_THREAD_POOL    (1U<<0)
+
+/*
+ * Is the default model supported? If THREAD_POOL is supported but DEFAULT_SPAWN
+ * is not supported, the custom thread spawner or manual thread method must be
+ * used.
+ */
+#define OSSL_THREAD_SUPPORT_FLAG_DEFAULT_SPAWN  (1U<<1)
+
+/* Returns zero or more of OSSL_THREAD_SUPPORT_FLAG_*. */
+uint32_t OSSL_get_thread_support_flags(void);
+```
+
+Since deciding on a reasonable number of threads is hardware-specific, a
+convenience function will also be provided to give a good ballpark figure:
+
+```c
+/*
+ * Get the recommended number of threads to be allocated to a thread pool.
+ *
+ * This is a hardware-specific value which approximates the number of
+ * concurrent threads of execution available on the machine.
+ *
+ * If thread pooling is disabled or not available, this may return 0 or may
+ * still return a valid value.
+ */
+uint32_t OSSL_get_recommended_threads(void);
+```
+
+Build Options
+-------------
+
+A build option `thread-pool`/`no-thread-pool` will be introduced which allows
+thread pool functionality to be compiled out. `no-thread-pool` implies
+`no-default-thread-pool`.
```

Comment:
I assume that the existing `no-threads` implies `no-thread-pool`?

Reply:
(Assuming the above assertion to be true) are there any (other) options that when enabled/disabled ought to disable thread{,-default}-pool?

Reply:
Yes to the assertion. IMO no other options should enforce disabling thread pool.